inject/src/main/java/io/micronaut/context/BeanResolutionTraceMode.java

```diff
@@ -0,0 +1,433 @@
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Allows enabling more verbose debugging on bean resolution.
+ */
+public enum BeanResolutionTraceMode {
+    /**
+     * No debug enabled (the default).
+     */
+    NONE,
+
+    /**
+     * When log mode is enabled output will be
+     * written to a logger named {@code io.micronaut.context.resolution} at DEBUG level.
+     */
+    LOG,
+
+    /**
+     * With standard out debug output will be written to {@link System#out} avoiding any log formatting.
+     */
+    STANDARD_OUT,
+
+    /**
+     * Interactive mode will leverage both {@link System#out} and {@link System#in}
+     * to allow pausing and resuming bean resolution and inspecting state.
+     */
+    INTERACTIVE;
+
+    static final Logger LOGGER = LoggerFactory.getLogger("io.micronaut.inject");
+    private static final String MODE_SYS_PROP = "micronaut.inject.trace.mode";
+    private static final String MODE_ENV_VAR = "MICRONAUT_INJECT_TRACE_MODE";
+    private static final String CLASSES_SYS_PROP = "micronaut.inject.trace";
+    private static final String CLASSES_ENV_VAR = "MICRONAUT_INJECT_TRACE";
+    private static final Set<String> INTERNAL_PACKAGES = Set.of(
+        "io.micronaut.context",
+        "io.micronaut.aop",
+        "io.micronaut.core.util",
+        "org.codehaus.groovy.vmplugin",
+        "java.util"
+    );
+
+    private static final String RIGHT_ARROW = AnsiColour.isSupported() ? " ➡️  " : " -> ";
+    private static final String RIGHT_ARROW_LOOP = AnsiColour.isSupported() ? " ↪️  " : "\\---> ";
+    private static final CharSequence START_TIME = "BeanResolutionTrace-start";
+
+    /**
+     * Obtains the default mode.
+     *
+     * @return The default mode
+     */
+    static BeanResolutionTraceMode getDefaultMode(Set<String> traceClasses) {
+        String mode = Optional
+            .ofNullable(System.getProperty(MODE_SYS_PROP))
+            .orElseGet(() -> System.getenv(MODE_ENV_VAR));
+        if (mode != null) {
+            return BeanResolutionTraceMode
+                .valueOf(NameUtils.environmentName(mode));
+        }
+        if (traceClasses.isEmpty()) {
+            return LOGGER.isTraceEnabled() ? BeanResolutionTraceMode.LOG : BeanResolutionTraceMode.NONE;
+        } else {
+            return LOGGER.isTraceEnabled() ? BeanResolutionTraceMode.LOG : BeanResolutionTraceMode.STANDARD_OUT;
+        }
+    }
+
+    static Set<String> getDefaultTraceClasses() {
+        String classes = Optional
+            .ofNullable(System.getProperty(CLASSES_SYS_PROP))
+            .orElseGet(() -> System.getenv(CLASSES_ENV_VAR));
+        if (classes != null) {
+            return Set.of(classes.split(","));
+        }
+        return Set.of();
+    }
+
+    void startTrace(
```

Comment:
Can you move all the tracing code into some other class, not to have it in a public enum?